## Segmenting and Clustering Neighborhoods in Toronto

In [1]:
import requests
from bs4 import BeautifulSoup
import csv
import json
import xml
import pandas as pd

#### Parsing the HTML of the entire Wikipedia page as a string

In [2]:
website_url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

#### Applying Beautiful Soup to make Soup using URL, then apply Prettify to view nested tags

In [3]:
from bs4 import BeautifulSoup

soup = BeautifulSoup(website_url,'lxml') # make Soup using URL

print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"4be8faea-0e68-4c1d-85a9-cdf7c27a3898","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":960187814,"wgRevisionId":960187814,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Communications in Ontario","Postal codes in Canada","Toron

#### The table is located under <"table class="wikitable sortable">

The HTML script of all the table contents is under the class wikitable sortable, where there are links to the neighborhood and postal code

Access tags by treating them as attributes when using soup.find() First locate the table, then parce out each neighborhood using 'tr'

In [4]:
neighborhood = soup.find('table', class_ = 'wikitable')
neighborhood_rows = neighborhood.find_all('tr')

#### Extract the columns 'Postcode', 'Borough', 'Neighbourhood' from the table

In [10]:
information = []
for row in neighborhood_rows:
    info = row.text.split('\n')[1:-1] # remove empty str (the first and last items)
    information.append(info)
    
information[0:20] #preview the first 20 rows

[['Postal Code', '', 'Borough', '', 'Neighborhood'],
 ['M1A', '', 'Not assigned', '', 'Not assigned'],
 ['M2A', '', 'Not assigned', '', 'Not assigned'],
 ['M3A', '', 'North York', '', 'Parkwoods'],
 ['M4A', '', 'North York', '', 'Victoria Village'],
 ['M5A', '', 'Downtown Toronto', '', 'Regent Park, Harbourfront'],
 ['M6A', '', 'North York', '', 'Lawrence Manor, Lawrence Heights'],
 ['M7A',
  '',
  'Downtown Toronto',
  '',
  "Queen's Park, Ontario Provincial Government"],
 ['M8A', '', 'Not assigned', '', 'Not assigned'],
 ['M9A', '', 'Etobicoke', '', 'Islington Avenue, Humber Valley Village'],
 ['M1B', '', 'Scarborough', '', 'Malvern, Rouge'],
 ['M2B', '', 'Not assigned', '', 'Not assigned'],
 ['M3B', '', 'North York', '', 'Don Mills'],
 ['M4B', '', 'East York', '', 'Parkview Hill, Woodbine Gardens'],
 ['M5B', '', 'Downtown Toronto', '', 'Garden District, Ryerson'],
 ['M6B', '', 'North York', '', 'Glencairn'],
 ['M7B', '', 'Not assigned', '', 'Not assigned'],
 ['M8B', '', 'Not assigne

#### Turn above information table into a Pandas dataframe

In [11]:
neighbor_df = pd.DataFrame(information[1:], columns=information[0])
# where information[1:] contains each row of neighborhoods
# and columns = information[0] gives the column names


neighbor_df.head(20)

Postal Code             Borough    \
0          M1A        Not assigned     
1          M2A        Not assigned     
2          M3A          North York     
3          M4A          North York     
4          M5A    Downtown Toronto     
5          M6A          North York     
6          M7A    Downtown Toronto     
7          M8A        Not assigned     
8          M9A           Etobicoke     
9          M1B         Scarborough     
10         M2B        Not assigned     
11         M3B          North York     
12         M4B           East York     
13         M5B    Downtown Toronto     
14         M6B          North York     
15         M7B        Not assigned     
16         M8B        Not assigned     
17         M9B           Etobicoke     
18         M1C         Scarborough     
19         M2C        Not assigned     

                                         Neighborhood  
0                                        Not assigned  
1                                        Not assigned  
2                                           Parkwoods  
3                                    Victoria Village  
4                           Regent Park, Harbourfront  
5                    Lawrence Manor, Lawrence Heights  
6         Queen's Park, Ontario Provincial Government  
7                                        Not assigned  
8             Islington Avenue, Humber Valley Village  
9                                      Malvern, Rouge  
10                                       Not assigned  
11                                          Don Mills  
12                    Parkview Hill, Woodbine Gardens  
13                           Garden District, Ryerson  
14                                          Glencairn  
15                                       Not assigned  
16                                       Not assigned  
17  West Deane Park, Princess Gardens, Martin Grov...  
18             Rouge Hill, Port Union, Highland Creek  
19                                       Not assigned

#### Data cleansing of the neighbor_df

- Get rid of Boroughs with "Not assigned"
- Combine Neighborhoods that share the same postcode
- For any Neighborhoods that are "not assigned," change to the Borough

#### Dropping any row where Borough is not assigned

In [12]:
neighbor_df = neighbor_df[neighbor_df.Borough != 'Not assigned']

neighbor_df.reset_index(drop=True, inplace=True)

neighbor_df.head(20)

Postal Code             Borough    \
0          M3A          North York     
1          M4A          North York     
2          M5A    Downtown Toronto     
3          M6A          North York     
4          M7A    Downtown Toronto     
5          M9A           Etobicoke     
6          M1B         Scarborough     
7          M3B          North York     
8          M4B           East York     
9          M5B    Downtown Toronto     
10         M6B          North York     
11         M9B           Etobicoke     
12         M1C         Scarborough     
13         M3C          North York     
14         M4C           East York     
15         M5C    Downtown Toronto     
16         M6C                York     
17         M9C           Etobicoke     
18         M1E         Scarborough     
19         M4E        East Toronto     

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4         Queen's Park, Ontario Provincial Government  
5             Islington Avenue, Humber Valley Village  
6                                      Malvern, Rouge  
7                                           Don Mills  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...  
12             Rouge Hill, Port Union, Highland Creek  
13                                          Don Mills  
14                                   Woodbine Heights  
15                                     St. James Town  
16                                 Humewood-Cedarvale  
17  Eringate, Bloordale Gardens, Old Burnhamthorpe...  
18                  Guildwood, Morningside, West Hill  
19                                        The Beaches

#### Combine neighborhoods with the same postcode

In [14]:
grouped = neighbor_df.groupby(['Postal Code']) # group by Postcode


# combine the neighborhoods grouped by postcode and into a new df
neighborhood_grouped = grouped['Neighborhood'].apply(lambda x: x.sum()) 
# adds spaces and commas between neighborhoods
neighborhood_grouped = grouped['Neighborhood'].apply(lambda x: "%s" % ', '.join(x))
# matches a borough to each postcode
borough_grouped = grouped['Borough'].apply(lambda x: set(x).pop())
# turn borough_grouped and neighborhood_grouped into dataframes
borough = borough_grouped.to_frame()
neighborhood = neighborhood_grouped.to_frame()
#combine the dataframe borough and the dataframe neighborhood into one dataframe
grouped_final = borough.merge(neighborhood, on="Postal Code")

grouped_final

Borough                                       Neighborhood
Postal Code                                                                
M1B          Scarborough                                     Malvern, Rouge
M1C          Scarborough             Rouge Hill, Port Union, Highland Creek
M1E          Scarborough                  Guildwood, Morningside, West Hill
M1G          Scarborough                                             Woburn
M1H          Scarborough                                          Cedarbrae
...                  ...                                                ...
M9N                 York                                             Weston
M9P            Etobicoke                                          Westmount
M9R            Etobicoke  Kingsview Village, St. Phillips, Martin Grove ...
M9V            Etobicoke  South Steeles, Silverstone, Humbergate, Jamest...
M9W            Etobicoke                Northwest, West Humber - Clairville

[103 rows x 2 columns]

## Attach longitude and latitude to the postcode, borough, and neighborhood table

Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

Given that the Geocoder Python package can be very unreliable, in case you are not able to get the geographical coordinates of the neighborhoods using the Geocoder package, here is a link to a csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data

In [15]:
geospatial_data = pd.read_csv('Geospatial_Coordinates.csv')

geospatial_data.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

#### Change the column name of Postal Code to Postcode

In [16]:
geospatial_data = geospatial_data.rename(columns={geospatial_data.columns[0]: "Postal Code" })

In [17]:
full_table = grouped_final.merge(geospatial_data, on = 'Postal Code')

full_table.head()

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

## Explore and cluster the neighborhoods of Toronto

In [18]:
#Installing some necessary packages

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 

!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library

print('Folium installed')
print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Folium installed
Libraries imported.


In [19]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

#### Let's see how many boroughs and neighborhoods are in Toronto

In [20]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(full_table['Borough'].unique()),
        full_table['Neighborhood'].shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


#### Get the longitude/latitude of Toronto using geolocator

In [21]:
address = 'Toronto, Ontario, Canada'

geolocator = Nominatim(user_agent="TO_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Create a map of Toronto with neighborhoods superimposed on top

In [22]:
# create a map of Toronto
map_toronto = folium.Map(location = [latitude, longitude], zoom_start = 10)

#add neighborhood markers to the Toronto map
for lat, long, bor, neigh in zip(full_table['Latitude'], full_table['Longitude'], 
                                 full_table['Borough'], full_table['Neighborhood']):
    label = '{}, {}'.format(neigh, bor)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius = 7, 
        popup = label,
        color = 'red',
        fill = True,
        fill_color = 'white',
        fill_opacity = 0.7,
        parse_html = False).add_to(map_toronto)
        
map_toronto

The map above shows all 103 neighborhoods in Toronto, using Folium map and a circle marker in the colors of the Canadian flag. Hovering over the markers, a label shows the neighborhood and the borough names.

#### Next, let's cluster by borough by first getting the geographic coordinates of each borough, then visualize via Folium

In [23]:
list_boroughs = full_table['Borough'].unique()
list_boroughs

array(['Scarborough', 'North York', 'East York', 'East Toronto',
       'Central Toronto', 'Downtown Toronto', 'York', 'West Toronto',
       'Mississauga', 'Etobicoke'], dtype=object)

In [24]:
def borough_loc(list_of_places):
    for place in list_of_places:
        address = (place + ", Ontario, Canada")
        geolocator = Nominatim(user_agent="TO_explorer")
        location = geolocator.geocode(address)
        latitude = location.latitude
        longitude = location.longitude
        print('{''}, {}, {},'.format(place,latitude,longitude))

borough_loc(list_boroughs)

Scarborough, 43.773077, -79.257774,
North York, 43.7543263, -79.44911696639593,
East York, 43.699971000000005, -79.33251996261595,
East Toronto, 43.72178945, -79.37402706301704,
Central Toronto, 43.6534817, -79.3839347,
Downtown Toronto, 43.6563221, -79.3809161,
York, 44.0007518, -79.4372217,
West Toronto, 43.6534817, -79.3839347,
Mississauga, 43.590338, -79.645729,
Etobicoke, 43.6435559, -79.5656326,


In [25]:
import numpy as np

boroughs = ['Scarborough', 43.773077, -79.257774,
'North York', 43.7708175, -79.4132998,
'East York', 43.6913391, -79.3278212,
'East Toronto', 43.653963, -79.387207,
'Central Toronto', 43.653963, -79.387207,
'Downtown Toronto', 43.655115, -79.380219,
'York', 44.0007518, -79.4372217,
'West Toronto', 43.653963, -79.387207,
"Queen's Park", 43.6599803, -79.3903686,
'Mississauga', 43.590338, -79.645729,
'Etobicoke', 43.6435559, -79.5656326]

boroughs_df = pd.DataFrame(np.array(boroughs).reshape(11,3), columns = ["Borough","Latitude","Longitude"])

boroughs_df

Borough    Latitude    Longitude
0        Scarborough   43.773077   -79.257774
1         North York  43.7708175  -79.4132998
2          East York  43.6913391  -79.3278212
3       East Toronto   43.653963   -79.387207
4    Central Toronto   43.653963   -79.387207
5   Downtown Toronto   43.655115   -79.380219
6               York  44.0007518  -79.4372217
7       West Toronto   43.653963   -79.387207
8       Queen's Park  43.6599803  -79.3903686
9        Mississauga   43.590338   -79.645729
10         Etobicoke  43.6435559  -79.5656326

In [26]:
boroughs_df.dtypes

boroughs_df["Latitude"] = pd.to_numeric(boroughs_df["Latitude"])
boroughs_df["Longitude"] = pd.to_numeric(boroughs_df["Longitude"])

In [27]:
# create a map of Toronto
map_toronto_boroughs = folium.Map(location = [43.653963, -79.387207], zoom_start = 10)

#add neighborhood markers to the Toronto map
for lat, long, bor in zip(boroughs_df['Latitude'], boroughs_df['Longitude'], 
                                 boroughs_df['Borough']):
    label = '{}'.format(bor)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius = 7, 
        popup = label,
        color = 'red',
        fill = True,
        fill_color = 'white',
        fill_opacity = 0.7,
        parse_html = False).add_to(map_toronto_boroughs)
        
map_toronto_boroughs

#### Want to explore the various neighborhoods within the boroughs. Let's see which borough has the most neighborhoods, then focus on a particularly populated borough

In [28]:
full_table['Borough'].value_counts()

North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East York            5
York                 5
East Toronto         5
Mississauga          1
Name: Borough, dtype: int64

It looks like North York has the most neighborhoods, with 24, followed by Downtown Toronto and Scarborough. 
I'm curious about the urban part of Toronto, so let's explore Downtown Toronto.

In [29]:
# The geographical coordinates for Downtown Toronto
DT_lat = boroughs_df.iloc[5,1]
DT_long = boroughs_df.iloc[5,2]

# The dataframe that contains all the Downtown Toronto neighborhoods
DT_df = full_table[full_table['Borough'] == 'Downtown Toronto'].reset_index(drop = True)

# create map of Downtown Toronto neighborhoods using latitude and longitude values
map_DT = folium.Map(location=[DT_lat, DT_long], zoom_start=12.4)

# add markers to map
for lat, lng, label in zip(DT_df['Latitude'], DT_df['Longitude'], DT_df['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='red',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_DT)  
    
map_DT

#### Let's play with some Foursquare data!

In [31]:
# Define Foursquare Credentials and Versions

CLIENT_ID = 'FAG31ZTWBFHKAYEC1QAMIQMCS3POFL1NSG5OPKILTUMBELJQ'
CLIENT_SECRET = 'JEKBDUCYSYZU2RH0XDOPFE5XXFGCYZGEZ5K43IWJZZXFL5HS'
VERSION = '20200620' # Foursquare API version

#### Let's explore the first neighborhood in our dataframe

Get the first neighborhood's name

In [32]:
DT_df.loc[0, 'Neighborhood']

'Rosedale'

Fetching Rosedale's longitude and latitude

In [33]:
neighborhood_latitude = DT_df.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = DT_df.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = DT_df.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Rosedale are 43.6795626, -79.37752940000001.


#### Now I will get the top 100 venues in Rosedale within a 500 meter radius.

First, I will create the GET request

In [34]:
LIMIT = 100

radius = 500

url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    lat, 
    lng, 
    VERSION, 
    radius, 
    LIMIT)

#### Send GET request and examine output

In [37]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ef8e11c07a2fd0bac92f0d8'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Bay Street Corridor',
  'headerFullLocation': 'Bay Street Corridor, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 33,
  'suggestedBounds': {'ne': {'lat': 43.6668015045, 'lng': -79.38328496866619},
   'sw': {'lat': 43.657801495499996, 'lng': -79.3957026313338}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b9d206bf964a520e69136e3',
       'name': "Queen's Park",
       'location': {'address': 'University Ave.',
        'crossStreet': 'at Wellesley Ave.',
        'lat': 43.66394609897775,
        'lng': -79.39217952520835,
        'labeledL

In [38]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [39]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

/Users/mypc/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name           categories        lat        lng
0                Queen's Park                 Park  43.663946 -79.392180
1                    Mercatto   Italian Restaurant  43.660391 -79.387664
2              Neo Coffee Bar          Coffee Shop  43.660140 -79.385870
3                   Sushi Box     Sushi Restaurant  43.662960 -79.386580
4                Central YMCA  Distribution Center  43.663083 -79.385025
5               Coffee Public          Coffee Shop  43.660763 -79.386184
6               T-Swirl Crepe             Creperie  43.663452 -79.384125
7                    Bar Volo             Beer Bar  43.665462 -79.385692
8                 Tokyo Sushi     Sushi Restaurant  43.665885 -79.386977
9          The Yoga Sanctuary          Yoga Studio  43.661499 -79.383636
10     Curry's Art Store Ltd.  Arts & Crafts Store  43.662838 -79.383732
11           Burrito Bandidos        Burrito Place  43.662962 -79.383956
12               Como En Casa   Mexican Restaurant  43.665160 -79.384796
13                  401 Games           Hobby Shop  43.663623 -79.384037
14     Coach House Restaurant                Diner  43.664991 -79.384814
15         Hart House Theatre              Theater  43.663571 -79.394616
16  Flock Rotisserie + Greens  Fried Chicken Joint  43.659167 -79.389475
17                  Dollarama       Discount Store  43.660680 -79.384680
18                     Fran's                Diner  43.661255 -79.383893
19                  Starbucks          Coffee Shop  43.658204 -79.388998
20              Booster Juice        Smoothie Shop  43.658657 -79.388682
21         Aroma Espresso Bar                 Café  43.658928 -79.389903
22             RBC Royal Bank                 Bank  43.659336 -79.389773
23                     Subway       Sandwich Place  43.658970 -79.390874
24                Tim Hortons          Coffee Shop  43.661038 -79.393797
25                  Starbucks          Coffee Shop  43.660887 -79.393720
26             Hart House Gym                  Gym  43.664172 -79.394888
27                Tim Hortons          Coffee Shop  43.659415 -79.391221
28                       SUDS                  Bar  43.659880 -79.394712
29           Convocation Hall   College Auditorium  43.660828 -79.395245
30                Tim Hortons          Coffee Shop  43.658175 -79.390681
31                Tim Hortons          Coffee Shop  43.658906 -79.388696
32              Gallery Grill           Restaurant  43.663841 -79.394309

In [40]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

33 venues were returned by Foursquare.


In [42]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(
                CLIENT_ID, 
                CLIENT_SECRET, 
                lat, 
                lng, 
                VERSION, 
                radius, 
                LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [44]:
DT_venues = getNearbyVenues(names=DT_df['Neighborhood'],
                                   latitudes=DT_df['Latitude'],
                                   longitudes=DT_df['Longitude']
                                  )

Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Queen's Park, Ontario Provincial Government


In [45]:
#Snapshot
print(DT_venues.shape)
DT_venues.head()

(1224, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                     Rosedale              43.679563              -79.377529   
1                     Rosedale              43.679563              -79.377529   
2                     Rosedale              43.679563              -79.377529   
3                     Rosedale              43.679563              -79.377529   
4  St. James Town, Cabbagetown              43.667967              -79.367675   

                  Venue  Venue Latitude  Venue Longitude Venue Category  
0         Rosedale Park       43.682328       -79.378934     Playground  
1          Whitney Park       43.682036       -79.373788           Park  
2  Alex Murray Parkette       43.678300       -79.382773           Park  
3        Milkman's Lane       43.676352       -79.373842          Trail  
4           Cranberries       43.667843       -79.369407          Diner

In [46]:
#Detailed Snapshot
DT_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            59   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     68   
Christie                                                               16   
Church and Wellesley                                                   74   
Commerce Court, Victoria Hotel                                        100   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
Kensington Market, Chinatown, Grange Park                              59   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                               94   
Rosedale                                                                4   
St. James Town                                                         78   
St. James Town, Cabbagetown                                            46   
Stn A PO Boxes                                                         97   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             59   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      68   
Christie                                                                16   
Church and Wellesley                                                    74   
Commerce Court, Victoria Hotel                                         100   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               59   
Queen's Park, Ontario Provincial Government                             33   
Regent Park, Harbourfront                                               45   
Richmond, Adelaide, King                                                94   
Rosedale                                                                 4   
St. James Town                                                          78   
St. James Town, Cabbagetown                                             46   
Stn A PO Boxes                                                          97   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          35   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            59              59   
CN Tower, King and Spadina, Railway Lands, Harb...     16              16   
Central Bay Street                                     68              68   
Christie                                               16              16   
Church and Wellesley                                   74              74   
Commerce Court, Victoria Hotel                        100             100   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100             100   
Harbourfront East, Union 

#### Analyzing

In [47]:
# one hot encoding
DT_onehot = pd.get_dummies(DT_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
DT_onehot['Neighborhood'] = DT_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [DT_onehot.columns[-1]] + list(DT_onehot.columns[:-1])
DT_onehot = DT_onehot[fixed_columns]

DT_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  \
0            0                  0        0                   0   
1            0                  0        0                   0   
2            0                  0        0                   0   
3            0                  0        0                   0   
4            0                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0                 0          0           0           0       0     0    0   
1                 0          0           0           0       0     0    0   
2                 0          0           0           0       0     0    0   
3                 0          0           0           0       0     0    0   
4                 0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0           0                   0       0              0          0         0   
1           0                   0       0              0          0         0   
2           0                   0       0              0          0         0   
3           0                   0       0              0          0         0   
4           0                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Butcher  Café  Camera Store  Candy Store  \
0             0              0        0     0             0            0   
1             0              0        0     0             0            0   
2             0              0        0     0             0            0   
3             0              0        0     0             0            0   
4             0              0        0     0             0            0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                     0            0                   0               0   
1                     0            0                   0               0   
2                     0            0                   0               0   
3                     0            0       

In [48]:
DT_onehot.shape

(1224, 206)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [49]:
DT_grouped = DT_onehot.groupby('Neighborhood').mean().reset_index()
DT_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
2                                  Central Bay Street     0.014706   
3                                            Christie     0.000000   
4                                Church and Wellesley     0.027027   
5                      Commerce Court, Victoria Hotel     0.000000   
6              First Canadian Place, Underground city     0.000000   
7                            Garden District, Ryerson     0.000000   
8   Harbourfront East, Union Station, Toronto Islands     0.000000   
9           Kensington Market, Chinatown, Grange Park     0.000000   
10        Queen's Park, Ontario Provincial Government     0.030303   
11                          Regent Park, Harbourfront     0.022222   
12                           Richmond, Adelaide, King     0.000000   
13                                           Rosedale     0.000000   
14                                     St. James Town     0.000000   
15                        St. James Town, Cabbagetown     0.000000   
16                                     Stn A PO Boxes     0.010309   
17           Toronto Dominion Centre, Design Exchange     0.000000   
18                     University of Toronto, Harbord     0.028571   

    Afghan Restaurant  Airport  Airport Food Court  Airport Lounge  \
0            0.000000   0.0000              0.0000           0.000   
1            0.000000   0.0625              0.0625           0.125   
2            0.000000   0.0000              0.0000           0.000   
3            0.000000   0.0000              0.0000           0.000   
4            0.013514   0.0000              0.0000           0.000   
5            0.000000   0.0000              0.0000           0.000   
6            0.000000   0.0000              0.0000           0.000   
7            0.000000   0.0000              0.0000           0.000   
8            0.000000   0.0000              0.0000           0.000   
9            0.000000   0.0000              0.0000           0.000   
10           0.000000   0.0000              0.0000           0.000   
11           0.000000   0.0000              0.0000           0.000   
12           0.000000   0.0000              0.0000           0.000   
13           0.000000   0.0000              0.0000           0.000   
14           0.000000   0.0000              0.0000           0.000   
15           0.000000   0.0000              0.0000           0.000   
16           0.000000   0.0000              0.0000           0.000   
17           0.000000   0.0000              0.0000           0.000   
18           0.000000   0.0000              0.0000           0.000   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0            0.0000             0.000             0.000000      0.000000   
1            0.1875             0.125             0.000000      0.000000   
2            0.0000             0.000             0.000000      0.000000   
3            0.0000             0.000             0.000000      0.000000   
4            0.0000             0.000             0.013514      0.000000   
5            0.0000             0.000             0.040000      0.000000   
6            0.0000             0.000             0.030000      0.000000   
7            0.0000             0.000             0.000000      0.000000   
8            0.0000             0.000             0.000000      0.000000   
9            0.0000             0.000             0.000000      0.000000   
10           0.0000             0.000             0.000000      0.000000   
11           0.0000             0.000             0.000000      0.022222   
12           0.0000             0.000             0.010638      0.000000   
13           0.0000             0.000             0.000000      0.000000   
14           0.0000             0.000             0.038462      0.000000   
15           0.0000    

#### Let's print each neighborhood along with the top 5 most common venues

In [50]:
num_top_venues = 5

for hood in DT_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = DT_grouped[DT_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
          venue  freq
0   Coffee Shop  0.08
1  Cocktail Bar  0.05
2    Restaurant  0.03
3      Beer Bar  0.03
4   Cheese Shop  0.03


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                venue  freq
0     Airport Service  0.19
1      Airport Lounge  0.12
2    Airport Terminal  0.12
3             Airport  0.06
4  Airport Food Court  0.06


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.16
1   Italian Restaurant  0.07
2       Sandwich Place  0.07
3  Japanese Restaurant  0.06
4                 Café  0.04


----Christie----
           venue  freq
0  Grocery Store  0.25
1           Café  0.19
2           Park  0.12
3     Restaurant  0.06
4          Diner  0.06


----Church and Wellesley----
                 venue  freq
0          Coffee Shop  0.08
1     Sushi Restaurant  0.07
2  Japanese Restaurant  0.05
3              Gay Bar  0.04
4           Restaurant  0.0

#### Some insights from the top venue per neighborhood analysis:

No surprise that in the Island Airport, Railway Lands neighborhoods, the top venues are airport lounges, airport service, airport terminal, plane, boat or ferry. These neighborhoods seem to be mostly have commuter related venues.

The top venues in Chinatown and Kensington Market are food establishments: Vietnamese and vegan restaurants, bars, bakeries, and cafes.

Church and Wellesley are neighborhoods that also have lots of food options, specifically Japanese cuisine.

Christie is the neighborhood to go to for eating in, which a high concentration of grocery and convenience stores. Eating in means you have more money to spend on nightclubs!

Craving caffeine and an afternoon snack? Central Bay Street is the neighborhood for you! There's many coffee shops and cafes and even a few burger joints.
Put the top venues information into a dataframe


#### First, write a function to sort the venues in descending order

In [51]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

#### Create a new dataframe and display the top 10 venues

In [52]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = DT_grouped['Neighborhood']

for ind in np.arange(DT_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(DT_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood 1st Most Common Venue  \
0                                         Berczy Park           Coffee Shop   
1   CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
2                                  Central Bay Street           Coffee Shop   
3                                            Christie         Grocery Store   
4                                Church and Wellesley           Coffee Shop   
5                      Commerce Court, Victoria Hotel           Coffee Shop   
6              First Canadian Place, Underground city           Coffee Shop   
7                            Garden District, Ryerson        Clothing Store   
8   Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
9           Kensington Market, Chinatown, Grange Park                  Café   
10        Queen's Park, Ontario Provincial Government           Coffee Shop   
11                          Regent Park, Harbourfront           Coffee Shop   
12                           Richmond, Adelaide, King           Coffee Shop   
13                                           Rosedale                  Park   
14                                     St. James Town           Coffee Shop   
15                        St. James Town, Cabbagetown           Coffee Shop   
16                                     Stn A PO Boxes           Coffee Shop   
17           Toronto Dominion Centre, Design Exchange           Coffee Shop   
18                     University of Toronto, Harbord                  Café   

   2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0           Cocktail Bar           Cheese Shop                 Bakery   
1         Airport Lounge      Airport Terminal       Sculpture Garden   
2     Italian Restaurant        Sandwich Place    Japanese Restaurant   
3                   Café                  Park             Restaurant   
4       Sushi Restaurant   Japanese Restaurant             Restaurant   
5             Restaurant                  Café                  Hotel   
6                   Café            Restaurant                  Hotel   
7            Coffee Shop        Cosmetics Shop        Bubble Tea Shop   
8               Aquarium                 Hotel                   Café   
9                 Bakery    Mexican Restaurant  Vietnamese Restaurant   
10      Sushi Restaurant                 Diner            Yoga Studio   
11                   Pub                  Park                 Bakery   
12            Restaurant                  Café                    Gym   
13            Playground                 Trail           Cupcake Shop   
14                  Café             Gastropub           Cocktail Bar   
15           Pizza Place                  Park             Restaurant   
16    Seafood Restaurant                  Café           Cocktail Bar   
17                 Hotel                  Café             Restaurant   
18            Restaurant                   Bar     Italian Restaurant   

   5th Most Common Venue          6th Most Common Venue  \
0     Seafood Restaurant                       Beer Bar   
1        Harbor / Marina            Rental Car Location   
2                   Café               Department Store   
3            Coffee Shop                      Nightclub   
4                Gay Bar                    Yoga Studio   
5                    Gym            American Restaurant   
6                    Gym             Seafood Restaurant   
7                   Café      Middle Eastern Restaurant   
8         Scenic Lookout                     Restaurant   
9            Coffee Shop  Vegetarian / Vegan Restaurant   
10    College Auditorium                  Smoothie Shop   
11        Breakfast Spot                        Theater   
12                 Hotel                Thai Restaurant   
13            Donut Shop               Doner Restaurant   
14   American Restaurant                     Restaurant   
15                   Pub                           Café

#### Cluster the Neighborhoods of Downtown Toronto

In [53]:
# run K-means to cluster the neighborhoods into 5 clusters

# set number of clusters
kclusters = 5

DT_grouped_clustering = DT_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(DT_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([2, 3, 2, 4, 2, 2, 2, 2, 2, 2], dtype=int32)

#### Create a new dataframe that inclues the cluster as well as the top 10 venues

In [54]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster', kmeans.labels_)

DT_merge = DT_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
DT_merge = DT_merge.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

DT_merge.head() # check the last columns!

Postal Code           Borough                 Neighborhood   Latitude  \
0         M4W  Downtown Toronto                     Rosedale  43.679563   
1         M4X  Downtown Toronto  St. James Town, Cabbagetown  43.667967   
2         M4Y  Downtown Toronto         Church and Wellesley  43.665860   
3         M5A  Downtown Toronto    Regent Park, Harbourfront  43.654260   
4         M5B  Downtown Toronto     Garden District, Ryerson  43.657162   

   Longitude  Cluster 1st Most Common Venue 2nd Most Common Venue  \
0 -79.377529        1                  Park            Playground   
1 -79.367675        2           Coffee Shop           Pizza Place   
2 -79.383160        2           Coffee Shop      Sushi Restaurant   
3 -79.360636        0           Coffee Shop                   Pub   
4 -79.378937        2        Clothing Store           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Trail          Cupcake Shop            Donut Shop   
1                  Park            Restaurant                   Pub   
2   Japanese Restaurant            Restaurant               Gay Bar   
3                  Park                Bakery        Breakfast Spot   
4        Cosmetics Shop       Bubble Tea Shop                  Café   

       6th Most Common Venue     7th Most Common Venue 8th Most Common Venue  \
0           Doner Restaurant                   Dog Run   Distribution Center   
1                       Café                    Bakery             Pet Store   
2                Yoga Studio  Mediterranean Restaurant                 Hotel   
3                    Theater                      Café  Gym / Fitness Center   
4  Middle Eastern Restaurant       Japanese Restaurant             Bookstore   

  9th Most Common Venue 10th Most Common Venue  
0        Discount Store                  Diner  
1    Chinese Restaurant     Italian Restaurant  
2           Men's Store        Bubble Tea Shop  
3           Event Space  Performing Arts Venue  
4      Ramen Restaurant          Movie Theater

#### Let's Examine Clusters

Cluster 1

In [55]:
DT_merge.loc[DT_merge['Cluster'] == 0, DT_merge.columns[[2] + list(range(5, DT_merge.shape[1]))]]

Neighborhood  Cluster  \
3                     Regent Park, Harbourfront        0   
18  Queen's Park, Ontario Provincial Government        0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
3            Coffee Shop                   Pub                  Park   
18           Coffee Shop      Sushi Restaurant                 Diner   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
3                 Bakery        Breakfast Spot               Theater   
18           Yoga Studio    College Auditorium         Smoothie Shop   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
3                   Café  Gym / Fitness Center           Event Space   
18              Beer Bar        Sandwich Place            Restaurant   

   10th Most Common Venue  
3   Performing Arts Venue  
18          Burrito Place

Cluster 2

In [56]:
DT_merge.loc[DT_merge['Cluster'] == 1, DT_merge.columns[[2] + list(range(5, DT_merge.shape[1]))]]

Neighborhood  Cluster 1st Most Common Venue 2nd Most Common Venue  \
0     Rosedale        1                  Park            Playground   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Trail          Cupcake Shop            Donut Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0      Doner Restaurant               Dog Run   Distribution Center   

  9th Most Common Venue 10th Most Common Venue  
0        Discount Store                  Diner

Cluster 3

In [57]:
DT_merge.loc[DT_merge['Cluster'] == 2, DT_merge.columns[[2] + list(range(5, DT_merge.shape[1]))]]

Neighborhood  Cluster  \
1                         St. James Town, Cabbagetown        2   
2                                Church and Wellesley        2   
4                            Garden District, Ryerson        2   
5                                      St. James Town        2   
6                                         Berczy Park        2   
7                                  Central Bay Street        2   
8                            Richmond, Adelaide, King        2   
9   Harbourfront East, Union Station, Toronto Islands        2   
10           Toronto Dominion Centre, Design Exchange        2   
11                     Commerce Court, Victoria Hotel        2   
12                     University of Toronto, Harbord        2   
13          Kensington Market, Chinatown, Grange Park        2   
15                                     Stn A PO Boxes        2   
16             First Canadian Place, Underground city        2   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
1            Coffee Shop           Pizza Place                  Park   
2            Coffee Shop      Sushi Restaurant   Japanese Restaurant   
4         Clothing Store           Coffee Shop        Cosmetics Shop   
5            Coffee Shop                  Café             Gastropub   
6            Coffee Shop          Cocktail Bar           Cheese Shop   
7            Coffee Shop    Italian Restaurant        Sandwich Place   
8            Coffee Shop            Restaurant                  Café   
9            Coffee Shop              Aquarium                 Hotel   
10           Coffee Shop                 Hotel                  Café   
11           Coffee Shop            Restaurant                  Café   
12                  Café            Restaurant                   Bar   
13                  Café                Bakery    Mexican Restaurant   
15           Coffee Shop    Seafood Restaurant                  Café   
16           Coffee Shop                  Café            Restaurant   

    4th Most Common Venue 5th Most Common Venue  \
1              Restaurant                   Pub   
2              Restaurant               Gay Bar   
4         Bubble Tea Shop                  Café   
5            Cocktail Bar   American Restaurant   
6                  Bakery    Seafood Restaurant   
7     Japanese Restaurant                  Café   
8                     Gym                 Hotel   
9                    Café        Scenic Lookout   
10             Restaurant   American Restaurant   
11                  Hotel                   Gym   
12     Italian Restaurant   Japanese Restaurant   
13  Vietnamese Restaurant           Coffee Shop   
15           Cocktail Bar            Restaurant   
16                  Hotel                   Gym   

            6th Most Common Venue     7th Most Common Venue  \
1                            Café                    Bakery   
2                     Yoga Studio  Mediterranean Restaurant   
4       Middle Eastern Restaurant       Japanese Restaurant   
5                      Restaurant                       Gym   
6                        Beer Bar                Restaurant   
7                Department Store               Salad Place   
8                 Thai Restaurant             Deli / Bodega   
9                      Restaurant        Italian Restaurant   
10                    Salad Place        Seafood Restaurant   
11            American Restaurant        Italian Restaurant   
12                      Bookstore                    Bakery   
13  Vegetarian / Vegan Restaurant                      Park   
15            Japanese Restaurant        Italian Restaurant   
16             Seafood Restaurant       Japanese Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1              Pet Store    Chinese Restaurant     Italian Restaurant  
2                  Hotel           Men's Store        Bubble Tea Shop  
4              Bookstore      Ramen Resta

Cluster 4

In [58]:
DT_merge.loc[DT_merge['Cluster'] == 3, DT_merge.columns[[2] + list(range(5, DT_merge.shape[1]))]]

Neighborhood  Cluster  \
14  CN Tower, King and Spadina, Railway Lands, Har...        3   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
14       Airport Service        Airport Lounge      Airport Terminal   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
14      Sculpture Garden       Harbor / Marina   Rental Car Location   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
14                 Plane                   Bar         Boat or Ferry   

   10th Most Common Venue  
14     Airport Food Court

Cluster 5

In [59]:
DT_merge.loc[DT_merge['Cluster'] == 4, DT_merge.columns[[2] + list(range(5, DT_merge.shape[1]))]]

Neighborhood  Cluster 1st Most Common Venue 2nd Most Common Venue  \
17     Christie        4         Grocery Store                  Café   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
17                  Park            Restaurant           Coffee Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
17             Nightclub                 Diner           Candy Store   

   9th Most Common Venue 10th Most Common Venue  
17    Italian Restaurant             Baby Store

### Clusters 3 was the biggest with 16, cluster 1 has only two neighbourhoods and clusters 2, 4 and 5 have just one neighbourhood each